### DISCUSSION TITLE
tnum font feature being mistakenly stripped from Noto Sans

### DISCUSSION THREAD

**Discussion by robinwhittleton** [harfbuzz/harfbuzz#4786] [Q&A] [answered]
Hi! I’ve got an odd problem that’s probably arising from me doing something wrong, but I can’t work out what. [I reported this earlier to subset-font](https://github.com/papandreou/subset-font/issues/24) (which uses harfbuzzjs at its core), but @papandreou also replicated the problem with hb-subset so the assumption is that it’s a bug either in harfbuzz or in the font.

I’ve made a Noto Sans font that replicates the issue: [NotoSansLatin-Regular.woff.zip](https://github.com/user-attachments/files/16051810/NotoSansLatin-Regular.woff.zip). This was downloaded from Google Fonts, then run through [fonttools](https://github.com/fonttools/fonttools) to strip most font features and save it as a woff, leaving it with `liga`, `tnum`, `zero` and `kern`.

The next step in our build chain is using [subset-font](https://github.com/papandreou/subset-font) to pass in a set of required glyphs to subset to. This calls out to harfbuzzjs, and the returned font now just has `liga`, `zero` and `kern` features: the `tnum` feature has disappeared.

Like I said, this is replicable with hb-subset, even if `tnum` is explicitely requested to be retained. Here’s @papandreou’s console output:

```sh
$ ./build/util/hb-subset --layout-features+=tnum ../subset-font/testdata/NotoSansLatin-Regular.ttf --text=foobar0123456789 -o hbsubsetted.ttf
$ ttx -o - hbsubsetted.ttf
[...]
Dumping 'GPOS' table...
        <FeatureTag value="kern"/>
Dumping 'GSUB' table...
Dumping 'STAT' table...
        <FeatureTag value="liga"/>
```

So: problem with hb-subset? Or problem with the current Noto Sans that hb-subset is tripping over?

**Comment by papandreou** [comment]
I tried downloading the original Noto Sans font from [Google Fonts](https://fonts.google.com/), and it looks like `hb-subset` does preserve the `tnum` feature when passing `--layout-features=*`:

```
$ ./build/util/hb-subset NotoSans-VariableFont_wdth,wght.ttf --text=foo123 --layout-features=* -o origsubset.ttf
$ ttx -o - origsubset.ttf | grep FeatureTag
Dumping "origsubset.ttf" to "-"...
Dumping 'GlyphOrder' table...
Dumping 'head' table...
Dumping 'hhea' table...
Dumping 'maxp' table...
Dumping 'OS/2' table...
Dumping 'hmtx' table...
Dumping 'cmap' table...
Dumping 'prep' table...
Dumping 'loca' table...
Dumping 'glyf' table...
Dumping 'name' table...
Dumping 'post' table...
Dumping 'gasp' table...
Dumping 'GDEF' table...
Dumping 'GPOS' table...
        <FeatureTag value="kern"/>
Dumping 'GSUB' table...
        <FeatureTag value="aalt"/>
        <FeatureTag value="case"/>
        <FeatureTag value="dnom"/>
        <FeatureTag value="frac"/>
        <FeatureTag value="liga"/>
        <FeatureTag value="lnum"/>
        <FeatureTag value="numr"/>
        <FeatureTag value="onum"/>
        <FeatureTag value="ordn"/>
        <FeatureTag value="pnum"/>
        <FeatureTag value="sinf"/>
        <FeatureTag value="smcp"/>
        <FeatureTag value="subs"/>
        <FeatureTag value="sups"/>
        <FeatureTag value="tnum"/>
```

(For some reason, `--font-features=tnum`, `--font-features=TNUM`, `--font-features+=tnum` or `--font-features+=TNUM` don't do the trick, not sure why)

However, the above does _not_ work with the font you provided that had been through fonttools first.

**Comment by khaledhosny** [comment, marked as answer]
The input font already has tabular figures as the default. Although it has a `tnum` feature, it applies to unreachable glyphs (old style and proportional figures, but the font has no features that uses them, nor `cmap` mapping). Since you are using `--text`, hb-subset will remove these unreachable glyphs and the `tnum` feature will be empty and thus removed as well. If you use `--glyphs='*'`, the `tnum` feature will be kept when requested (either with `--layout-features+=tnum`, or `--layoutfeatures='*'`), but it would still be superfluous since it applies to unreachable glyphs.

**Comment by robinwhittleton** [comment]
Thanks all for the input. It looks like there’s two bugs at play here:

1. Something in fonttools / Noto Sans hiding the real issue
2. A bug in the new version of Noto we (IKEA) use where tabular numbers is not working correctly

I really appreciate you all taking the time to look at this, but as it’s not anything to do with harfbuzz I’ll close this now.
